File: src/main/java/com/github/liebharc/queryenrichment/Dependencies.java
package com.github.liebharc.queryenrichment;

import java.util.*;

/**
 * Utility class with some default dependencies.
 */
public class Dependencies {
    /** Singleton for no dependency */
    private static final NoDependency noDependency = new NoDependency();

    private Dependencies() {
        // Utility class
    }

    /** A step has no dependencies */
    public static Dependency noDependencies() {
        return noDependency;
    }

    /** A step requires all of the given attributes */
    public static Dependency requireAll(Attribute<?>... attributes) {
        return new RequireAll(Arrays.asList(attributes));
    }

    /** A step requires a single attribute */
    public static Dependency require(Attribute<?> attribute) {
        return new RequireAll(Collections.singletonList(attribute));
    }

    /** A step requires one of many attributes */
    public static Dependency requireOneOf(Attribute<?>... attributes) {
        return new RequireOneOf(Arrays.asList(attributes));
    }

    private static class RequireOneOf implements Dependency {

        private final Collection<Attribute<?>> attributes;

        private RequireOneOf(Collection<Attribute<?>> attributes) {
            this.attributes = attributes;
        }

        @Override
        public Collection<Attribute<?>> getMinimalRequiredAttributes(Collection<Attribute<?>> available) {
            if (this.isEmpty()) {
                return Collections.emptyList();
            }

            final Optional<Attribute<?>> any = attributes.stream().filter(available::contains).findAny();
            if (any.isPresent()) {
                return Collections.singletonList(any.get());
            } else {
                // We have no match at all, inform the caller about one of the dependencies as this is the minimum
                // we require
                return Collections.singletonList(attributes.iterator().next());
            }
        }

        @Override
        public boolean isEmpty() {
            return attributes.isEmpty();
        }

        @Override
        public boolean isOkay(Set<Attribute<?>> available) {
            if (this.isEmpty()) {
                return true;
            }

            return this.attributes.stream().anyMatch(available::contains);
        }
    }


    private static class RequireAll implements Dependency {

        private final Collection<Attribute<?>> attributes;

        private RequireAll(Collection<Attribute<?>> attributes) {
            this.attributes = attributes;
        }

        @Override
        public Collection<Attribute<?>> getMinimalRequiredAttributes(Collection<Attribute<?>> available) {
            return this.attributes;
        }

        @Override
        public boolean isEmpty() {
            return attributes.isEmpty();
        }

        @Override
        public boolean isOkay(Set<Attribute<?>> available) {
            return available.containsAll(this.attributes);
        }
    }

    private static class NoDependency implements Dependency {

        @Override
        public Collection<Attribute<?>> getMinimalRequiredAttributes(Collection<Attribute<?>> available) {
            return Collections.emptyList();
        }

        @Override
        public boolean isEmpty() {
            return true;
        }

        @Override
        public boolean isOkay(Set<Attribute<?>> constantAttributes) {
            return true;
        }
    }
}

File: src/main/java/com/github/liebharc/queryenrichment/Dependencies.kt
package com.github.liebharc.queryenrichment

import java.util.*

/**
 * Utility class with some default dependencies.
 */
object Dependencies {
    /** Singleton for no dependency  */
    private val noDependency = NoDependency()

    /** A step has no dependencies  */
    fun noDependencies(): Dependency {
        return noDependency
    }

    /** A step requires all of the given attributes  */
    fun requireAll(vararg attributes: Attribute<*>): Dependency {
        return RequireAll(Arrays.asList(*attributes))
    }

    /** A step requires a single attribute  */
    fun require(attribute: Attribute<*>): Dependency {
        return RequireAll(listOf(attribute))
    }

    /** A step requires one of many attributes  */
    fun requireOneOf(vararg attributes: Attribute<*>): Dependency {
        return RequireOneOf(Arrays.asList(*attributes))
    }

    /** A step with only optional dependencies  */
    fun optional(vararg attributes: Attribute<*>): Dependency {
        return OptionalDependency(Arrays.asList(*attributes))
    }

    private class RequireOneOf constructor(private val attributes: Collection<Attribute<*>>) : Dependency {

        override val isEmpty: Boolean
            get() = attributes.isEmpty()

        override fun getMinimalRequiredAttributes(
                selection: Map<Attribute<*>, ExecutableStep<*, *>>,
                available: Collection<Attribute<*>>): Collection<Attribute<*>> {
            if (this.isEmpty) {
                return emptyList()
            }

            val any = attributes.firstOrNull { available.contains(it) }
            return if (any != null) {
                listOf(any)
            } else {
                // We have no match at all, inform the caller about one of the dependencies as this is the minimum
                // we require
                listOf(attributes.iterator().next())
            }
        }

        override fun canBeConstant(attributes: Set<Attribute<*>>): Boolean {
            return if (this.isEmpty) {
                true
            } else {
                this.attributes.any { attributes.contains(it) }
            }

        }
    }


    private class RequireAll constructor(private val attributes: Collection<Attribute<*>>) : Dependency {

        override val isEmpty: Boolean
            get() = attributes.isEmpty()

        override fun getMinimalRequiredAttributes(
                selection: Map<Attribute<*>, ExecutableStep<*, *>>,
                available: Collection<Attribute<*>>): Collection<Attribute<*>> {
            return this.attributes
        }

        override fun canBeConstant(attributes: Set<Attribute<*>>): Boolean {
            return attributes.containsAll(this.attributes)
        }
    }

    private class NoDependency : Dependency {

        override val isEmpty: Boolean
            get() = true

        override fun getMinimalRequiredAttributes(
                selection: Map<Attribute<*>, ExecutableStep<*, *>>,
                available: Collection<Attribute<*>>): Collection<Attribute<*>> {
            return emptyList()
        }

        override fun canBeConstant(attributes: Set<Attribute<*>>): Boolean {
            return true
        }
    }

    private class OptionalDependency constructor(private val attributes: Collection<Attribute<*>>) : Dependency {

        override val isEmpty: Boolean
            get() = attributes.isEmpty()

        override fun getMinimalRequiredAttributes(
                selection: Map<Attribute<*>, ExecutableStep<*, *>>,
                available: Collection<Attribute<*>>): Collection<Attribute<*>> {
            return attributes.filter { atr -> attributes.contains(atr) && selection.containsKey(atr) }
        }

        override fun canBeConstant(attributes: Set<Attribute<*>>): Boolean {
            return false
        }
    }
}
